### ISSUE TITLE
Feature Request - Support using session scope-fixtures when using pytest.mark.parametrize 

### ISSUE THREAD

**Issue by amitwer** [pytest-dev/pytest#7312]
Hi, 
I've encountered a situation that I couldn't find a way to solve (in a clean way) with pytest current abilities, and thought it is general enough to include in the core project instead of just hacking something myself. 

Use case: 
I have a test that gets a single parameter and performs a scenario. Under one configuration - there are 3 relevant values, under another - 4 values are relevant). The problem is - the configuration is being determined in a session-scoped fixture, and is not accessible to me when defining the parameters. 

Currently, I can either use pytest_generate_test - which has a rather ugly syntax. Instead it would be much prettier if I could use something like: 

```
@pytest.fixture
def fixme(session_fixture):
 return [1,3,7] if session_fixture.config.is_premium else [1,4]

@pytest.mark.parametrize("x", parameters_fixture=my_fixture)
def test_me(x):
    print(amit)

```

Thanks

**Comment by RonnyPfannschmidt** [comment]
There is a issue about fixtures usable there, it's currently out on a near term Roadmap

Im on mobile so i can't search for it atm

**Comment by SalmonMode** [comment]
You can define all the tests, and then filter which ones you want to run, rather than providing a system for generating tests. E.g.:

```python
@pytest.fixture(scope="session", params=[(1, 3, 7), (1, 4)], ids=("premium", "standard")
def x(request):
    return request.param

def test_me(x):
    # stuff
```

And then you can just call it with `pytest -k "premium"`. You could also leverage marks instead of ids.

**Comment by amitwer** [comment]
that's an interesting idea, but sadly it's not as simple - just getting the data for making a decision is a multi-step process (e.g. - read a file, extrapolate data from it). For simple cases we use conftest to filter out tests which are not suitable, but going through such hidden piece of code is not something we'd like to do for a specific test logic.

**Comment by SalmonMode** [comment]
Sounds like you have a pretty complex file. I would recommend creating a custom data type and a parser where that logic is contained, so you can show tuck the logic away through abstraction but have a reference that goes straight to it. I'm imagining something like this:

```python
class Fruits(Enum):
    apple = Apple
    banana = Banana
    cherry = Cherry

fruits = []
with open("fruit_data.csv") as csvfile:
    reader = csv.DictReader(csvfile)
    for row in reader:
        fruits.append(Fruits[row["type"]](**row))

def fruit_id(fruit):
    fruit_repr = repr(fruit)
    if fruit.is_premium:  # would be pulled from the CSV and the fruit class would look for this arg during __init__
        return fruit_repr.replace("<", "<+", 1)
    return fruit_repr.replace("<", "<-", 1)

def pytest_collection_modifyitems(items):
    for item in items:
        if "<-" in item.nodeid:
            item.add_marker(pytest.mark.standard_fruit)
        elif "<+" in item.nodeid:
            item.add_marker(pytest.mark.premium_fruit)

@pytest.fixture(scope="session", params=fruit, ids=fruit_id)
def fruit(request):
    return request.param

def test_me(fruit):
    # stuff
```

The role of the DictReader can be replaced by a custom parser if you're not working with a CSV.

This would allow selecting based on marks, rather than test names.

I'm not sure if this approach is applicable to your scenario, but if you provide some more context, I might be able to come up with something more appropriate.

**Comment by amitwer** [comment]
I see what you are doing here, and while it can work, it's a lot of boilerplate code I was hoping to save.
Some context: I'm testing a product with multiple versions and flavors and has a client and a server. 
Now, I have a basic test which is applicable to all flavors and versions, but the parameters are different. 
e.g.: 
```python
def get_file_types():
    server = ServerFactory.server()
    client = ClientFactory.client()
    file_types = ['xlsx','pdf','zip']
    server_config = server.fetch_config() # Send REST request to the server
    client_config = client.read_config() # run a cmd on the client
    if client_config.OS is Os.WIN and client_config.os_version is OsVersion.Win7 and server_config.is_supporting_legacy_files:
        file_types.append['rtf']

def test_process_file(get_file_types()):
# do stuff
    
```

The idea is that creating the server & client objects (with the relevant factories) is an expensive operation I already do at a session fixture since most tests need it. What I currently do is to invoke the code used by the fixtures and pay the cost of duplicating the fixture's code, and I don't get to enjoy the pytest ecosystem. Each challenge I face can be solved using one hack or another ( I could build a cache and store the client and server objects according to my needs, for instance) but the motivation behind my request is that any solution I can tailor to my needs will be unnatural to the way the rest of the code is built (as it was built with pytest in mind)

**Comment by SalmonMode** [comment]
I'm not sure I follow, so I'm gonna propose something, and hopefully that'll highlight what I might be missing.

I'm not exactly sure what your situation is, but I find it's easier to leave server instantiation outside of the test logic, and treat it as infrastructure so everything in the test logic can just assume it is already there. Then it's as easy as using configs to point the tests to the right server. This applies even when you have different tests for each server.

That said, it looks like trying to compact the logic for parameterization is where the complexity becomes cumbersome, so that's where I'd say it's better to keep things simple. Off the top of my head, I'm thinking of something like this:

`tests/conftest.py` (top level conftest)

```python
def get_supported_file_type_params():
    supported_file_types = []
    for client_os_version in OS_VERSION:
        for server_os_version in OS_VERSION:
            file_types = ["xlsx", "pdf", "zip"]
            if server_os_version is OS_VERSION.WINDOWS7 and client_os_version.family is OS.WINDOWS:
                file_types.append("rtf")

            supported_file_types = {
                "client_os": client_os_version,
                "server_os_": server_os_version,
                "file_types": file_types,
            }
    return supported_file_types


def os_id_func(value):
    return "<OSData Client: {client_os} Server: {server_os} >"

@pytest.fixture(scope="session", params=get_supported_file_types(), ids=os_id_func)
def os_data(request):
    return request.param


@pytest.fixture(scope="session")
def client_os(os_data):
    return os_data["client_os"]


@pytest.fixture(scope="session")
def server_os(os_data):
    return os_data["server_os"]


@pytest.fixture(scope="session")
def file_types(os_data):
    return os_data["file_types"]


@pytest.fixture(scope="session")
def client(request):
    return Client(request.config.option.server_url)


def pytest_collection_modifyitems(items):
    for item in items:
        item_params = item.nodeid.split("[")[-1][:-1].split("-")
        
        os_data_repr = result = filter(lambda x: x.startswith("<OSData "), item_params)
        _, _, client_os, _, server_os, _ = os_data_repr.split()

        item.add_marker(f"client_os_{client_os.lower().replace(" ", "_")}")
        item.add_marker(f"client_os_family_{OS(client_os).family.lower().replace(" ", "_")}")
        item.add_marker(f"server_os_{server_os.lower().replace(" ", "_")}")
```

`tests/subpackage/test_something.py`

```python
@pytest.fixture
def process_file(client, file_types):
    client.do_something(file_types)


def test_process_file(process_file, client, file_types):
    assert client.get_latest_files() == file_types
```

With this, all tests exist, and it's just a matter of providing the configs necessary so the client can connect to the server properly, and then using the marks to select the tests based on the currently active infrastructure, e.g. `pytest -m "client_os_windows_7 and server_os_windows_7"`. Because the fixtures in the top-level conftest are session-scoped, they will only be executed once per parameter.

Keep in mind that fixtures are evaluated based on the perspective of the test that's about to run, so you can override `os_data` to be whatever you like for subsets of tests that don't need to apply to all permutations. So if none of the tests in `tests/subpackage` need to run when the client is Windows 7, you can define another `conftest.py` in there and override the `os_data` fixture with new params by just filtering out what is given by `get_supported_file_type_params` (so this may be best defined in a separate part of the package). Because of how pytest fixtures work, you only need to override that one fixture and everything else will be taken care of.

Also, this is _super_ generalized, and really just meant to showcase a few options/techniques. So it's just my attempt to throw stuff at the wall and see what sticks. I'm still not sure what your exact needs are, but with more context I can probably provide a better solution if this doesn't do it for you.

**Comment by amitwer** [comment]
I'm not sure I follow exactly the solution you offer, but I think it does help to pinpoint the missing detail: the logic I need is relevant for a single test, and there are several tests that need the same complex objects to decide on the relevant parameters. So on one hand, I don't want to do a lot of work in the test file, but on the other, I don't want to clutter contest with a method for each test.

**Comment by SalmonMode** [comment]
Hmmm I think I'm getting a clearer picture for what you need, but I'm not positive. Can you provide some more detail about the complex objects you mentioned and what the relevant parameters would be as a result of them?

Also, "parameters" is a kind of loaded term in this context, though, so just to clarify, it can mean one of:

1. a collection of inputs from which to generate multiple sets of tests, one for each input. FOr example, "chrome" and "firefox" being used to parameterize a `driver` fixture so that all tests that use `driver` will have two versions of themselves made so that they all run once with chrome, and then again with firefox.
2. A collection of data used to configure some entity. For example, `["xlsx", "pdf", "zip"]` being used to tell a server what files it should automatically be rejecting.

I'm sure I'm missing some definitions, and this distinction may seem arbitrary, but I'm mostly looking to distinguish between the first definition and everything else, as this can change the approach I'd take.

When you say "relevant parameters" are you referring to the 1st definition? or something else?

**Comment by amitwer** [comment]
I'm referring to the 1st definition. 
I'm unsure if I'm allowed sharing real examples from our code, so I'll try giving a close enough case: 
Assume we are monitoring file edits on the OS. There are several ways of editing the file, depending on the OS, architecture (x32, x64) and older product versions might be monitoring only some of those ways of edit, or supporting different OSes. Furthermore - Server config might mean we monitor only some of the edits. I have a single test for this feature
```python
@pytest.mark.client
@pytest.mark.parametrize("edit_method", [WRITE_OVER,ENCRYPT,DEL_AND_CREATE,REPLACE])
def test_file_edit_is_monitored(edit_method):
   _edit_file(edit_method,DUMMY_FILE_PATH)
   assert_that(self.server.get_edited_files()).contains(DUMMY_FILE_PATH)
``` 

Now, running this on win7x32 will fail on REPLACE, and if i'm in version lower than 1.5 ENCRYPT is not supported. 
getting the needed information CAN be gathered before collection, so instead of the list I have a method _get_relevant_params() or something with a better name that does the logic, but since it is called during collection phase, I need to support  odd cases - If I'm running a serverless flavor where   this test is irrelevant (I'm using -m not client), I won't have some of the data (e.g. server.version) so I keep adding exception handling just to avoid failing collection of such cases, and having similar, bu not identical code all over my test files

**Comment by SalmonMode** [comment]
Ah, gotcha! Thanks for clarifying and providing the extra context.

So in this case, if I understand correctly, I would define 4 separate tests, one for each edit method. You can mark and/or parameterize them individually according to the supported environments, and then have something external (either a human or a script) detect what the current environment is, and then know how to call pytest so that only the applicable tests are running (probably by using the marks). Having the logic and marks in place before the tests are run also avoids any runtime hiccups related to trying to diagnose what to run automatically (or at least decouples it from your test suite).

It may seem a little against DRY, but ultimately, you get something simple and descriptive (e.g. `test_write_over_is_monitored`), and KISS/DAMP are more beneficial than DRY in the long run.

**Comment by amitwer** [comment]
Yes, that's something we've considered doing as well when it was only this single case, and then we were hit by a flurry of such cases - a table with changed columns over versions (test filter column by value, when the table has 30+ columns), a new feature only partially working on win7, different user roles when working in a multi-tenant configuration (check role X can be created for ~10 roles) , and it seems to be a new case every day as we catch up feature coverage of a fairly complex product. DAMPening our tests to that extent would probably more than double the test portion of our code and will continue to bloat with each new version out in the field.

**Comment by amitwer** [comment]
@RonnyPfannschmidt  - A colleague of mine facing the same need (which is sensible, as we work on the same project) found this issue: https://github.com/pytest-dev/pytest/issues/349 
Could this be what you've referred to?

**Comment by RonnyPfannschmidt** [comment]
I think so, thanks for the lookup

**Comment by SalmonMode** [comment]
With parameterization, you'll effectively still have as many test cases. This approach just compartmentalizes the complexity a bit.

**Comment by amitwer** [comment]
I will have as many test cases, but the maintenance cost does not come from running the tests, but rather from fixing my mistakes(or adapting to a new need) in many places at once. 

Imagine I have the following code: 

```python
@pytest.mark.client
@pytest.mark.parametrize("edit_method", [WRITE_OVER,ENCRYPT,DEL_AND_CREATE,REPLACE])
def test_undo_after_reboot(edit_method,client): #client comes from a fixture
    original_content = read_file(DUMMY_FILE_PATH) #1
    _edit_file(edit_method,DUMMY_FILE_PATH) #2
    client.reboot() #3
    _undo_edit(DUMMY_FILE) #4
   assert_that(read_file(DUMMY_FILE_PATH)).is_equal_to(original_content) #5
```
Now, let's assume I duplicate this test in order to support the multiple configurations and now have 3 versions of it (x32 win7, old version & new version). Furthermore, let's assume that due to a security update we now need to explicitly turn on the "undo mode" after rebooting (It's a lame excuse, but this is what happens when I try to come up with examples that are concise enough) so now I need to add to my test (after line 3) something like "client.enable_undo()", only now I have to do that on 3 tests instead of 1. 
Sure, I can do something like extracting the test logic to a private function (or somehow push all actions except the assert to the fixture) but going back to the DAMP principle - It is less descriptive to anyone reading the test.

**Comment by SalmonMode** [comment]
Ahhhhhh, I see the problem.

My approach is to keep everything out of the test function itself except for a single assert, which is only a non-state-changing query.

I rely on multiple fixtures to perform the "arrange" steps, and "act" command, and limit my fixtures to doing (state-changing commands) and/or providing one thing at most. This helps with injecting dependencies into the flow higher up, as I can change any individual resource In my fixtures, and influence any step of the process. I can also pick and choose the steps I want, and more easily coordinate the "order of operations" so to speak of fixtures for a given test or collection of tests.

It also allows me to have very safe teardowns, because a fixture that throws an error likely won't actually change the state (because it failed to do what it wanted), and that means all the state-changing fixtures that ran before it will still try to teardown (also the logic for "undoing" a state change is right next to the logic that did that change). This also has the benefit of providing a nice readout when I run pytest with `--setup-plan` because it spits out every step and resource used for a test (and I find it's much easier to name fixtures when doing this).

I also try to handle parameterization in isolation from everything by doing it in its own fixture, as this makes manipulating that parameterization easy, and makes depending on that parameterization easy as well, since any test that can see that fixture can request it.

I also rely heavily on scope to segregate certain tests from others, but this also makes it super easy to provide a modified version of a specific fixture only for those scopes.

(I also use larger scopes like class to house multiple test functions to run on a given state that they all share the same setup for, which helps optimize a bit, but this isn't necessary. I'll set the fixtures to use the class scope for this example to demonstrate this, but the general solution will still work inside a class without the fixtures being scoped that way)

So if I restructure your test a bit with that in mind, I'd get this (assuming this is in a test file meant only for testing undo):

```python
@pytest.fixture(scope="class")
def file_original_content(os_data, file_path):
    # requests os_data to ensure it's impacted by parameterization
    return read_file(file_path)


@pytest.fixture(scope="class")
def edit_file(edit_method, file_path, os_data, file_original_contents):
    # requests os_data to ensure it's impacted by parameterization

    # requests file_original_contents to ensure the contents are grabbed before editing
    edit_file(edit_method, file_path)


@pytest.fixture(scope="class")
def reboot_client(edit_file, client):
    # requests edit_file because this needs to be done after editing the file
    # will be affected by parameterization because edit_file requests os_data
    client.reboot()


@pytest.fixture(scope="class", autouse=True)
def undo_edit(reboot_client, file_path):
    # requests reboot_client because this needs to be done after the client is rebooted

    # will be affected by parameterization because reboot_client effectively requests os_data

    # autouse is used as this is the last fixture to run in the dependency tree and this ensures all 
    # fixtures will run before the test functions begin their asserts. It also cuts out the need to
    # request the undo_edit fixture in the test.
    undo_edit(file_path)


class TestWriteOverEvent
    @pytest.fixture(scope="class", params=filter(lambda x: True, os_version_config_data_stuff))
    def os_data(self, request):
        return request.param

    def test_contents_are_restored(self, file_path, file_original_contents):
        assert read_file(file_path) == file_original_contents

    def test_server_saw_edit_event(self, server, file_path):
        # doesn't trigger a re-run of the fixtures because they were all class-scoped and autouse
       assert file_path in server.get_edited_files()
```

To add in the bit that enables undo, I would just modify the requests of `undo_edit` and add another fixture:

```python

@pytest.fixture(scope="class")
def enable_undo(reboot_client, client):
    # requests reboot_client because this needs to be done after the client is rebooted

    # will be affected by parameterization because reboot_client effectively requests os_data
    client.enable_undo()
    

@pytest.fixture(scope="class", autouse=True)
def undo_edit(reboot_client, enable_undo, file_path):
    # requests reboot_client because this needs to be done after the client is rebooted

    # requests enable_undo to ensure it only attempts to enable it after the client is rebooted

    # will be affected by parameterization because reboot_client effectively requests os_data

    # autouse is used as this is the last fixture to run in the dependency tree and this ensures all 
    # fixtures will run before the test functions begin their asserts. It also cuts out the need to
    # request the undo_edit fixture in the test.
    undo_edit(file_path)
```

A lot of these fixtures can be defined in a conftest.py file in a higher scope to avoid duplication.

**Comment by amitwer** [comment]
That's a very different way of writing the code. I can see how it might be useful, but changing to it would be quite a paradigm shift

**Comment by SalmonMode** [comment]
Absolutely. It'd be a big shift, but it's been _enormously_ helpful for me, and definitely helped keep things straightforward, fast, flexible, and safe (safe in regards to the teardowns)
